File: src/war3map/UnitControllers/BasicUnitController.lua
local ____exports = {}
local __TSTL_Point = require("war3map.Generic.Point")
local Point = __TSTL_Point.Point
local __TSTL_ExtensionFunctions = require("war3map.ExtensionFunctions")
local IsWalkable = __TSTL_ExtensionFunctions.IsWalkable
local __TSTL_Misc = require("war3map.Generic.Misc")
local rotateToPoint = __TSTL_Misc.rotateToPoint
____exports.BasicUnitController = {}
local BasicUnitController = ____exports.BasicUnitController
BasicUnitController.name = "BasicUnitController"
BasicUnitController.__index = BasicUnitController
BasicUnitController.prototype = {}
BasicUnitController.prototype.__index = BasicUnitController.prototype
BasicUnitController.prototype.constructor = BasicUnitController
function BasicUnitController.new(...)
    local self = setmetatable({}, BasicUnitController.prototype)
    self:____constructor(...)
    return self
end
function BasicUnitController.prototype.____constructor(self, targetUnit, heroUnit)
    self.currentDirection = bj_UNIT_FACING
    self.targetUnit = targetUnit
    self.walkAnimationIndex = heroUnit.walkAnimationIndex
    self.heroUnit = heroUnit
end
function BasicUnitController.prototype.moveUnit(self, toDirection)
    local speed = GetUnitMoveSpeed(self.targetUnit) / 100
    self.currentDirection = rotateToPoint(nil, self.currentDirection, toDirection, 180)
    local lastLoc = Point:fromLocationClean(GetUnitLoc(self.targetUnit))
    local currLoc = lastLoc:polarProject(speed, self.currentDirection)
    if IsWalkable(nil, currLoc.x, lastLoc.y) then
        SetUnitX(self.targetUnit, currLoc.x)
    end
    if IsWalkable(nil, lastLoc.x, currLoc.y) then
        SetUnitY(self.targetUnit, currLoc.y)
    end
    SetUnitFacing(self.targetUnit, self.currentDirection)
end
function BasicUnitController.prototype.getWalkAnimationIndex(self)
    return self.walkAnimationIndex
end
return ____exports

File: war3map/UnitControllers/BasicUnitController.ts
import {IUnitController} from "./IUnitController";
import {HeroUnit} from "./HeroUnit";
import {Point} from "../Generic/Point";
import {IsWalkable} from "../ExtensionFunctions";
import {rotateToPoint} from "../Generic/Misc";

export class BasicUnitController implements IUnitController {
    private readonly targetUnit: unit;
    private readonly walkAnimationIndex: number;
    private currentDirection: number = bj_UNIT_FACING;
    public heroUnit: HeroUnit;


    constructor(targetUnit: unit, heroUnit: HeroUnit) {
        this.targetUnit = targetUnit;
        this.walkAnimationIndex = heroUnit.walkAnimationIndex;
        this.heroUnit = heroUnit;
    }

    public moveUnit(toDirection: number) {
        let speed = GetUnitMoveSpeed(this.targetUnit) / 100;
        this.currentDirection = rotateToPoint(this.currentDirection, toDirection, 180);
        let lastLoc = Point.fromLocationClean(GetUnitLoc(this.targetUnit));
        let currLoc = lastLoc.polarProject(speed, this.currentDirection);
        if (IsWalkable(currLoc.x, lastLoc.y)) {
            SetUnitX(this.targetUnit, currLoc.x);
        }
        if (IsWalkable(lastLoc.x, currLoc.y)) {
            SetUnitY(this.targetUnit, currLoc.y);
        }
        SetUnitFacing(this.targetUnit, this.currentDirection);
    }

    public getWalkAnimationIndex(): number {
        return this.walkAnimationIndex;
    }
}